### ISSUE TITLE
PerMonitorV2 DPI mode support

### ISSUE THREAD

**Issue by danergo** [kirsan31/winforms-datavisualization#26]
First of all: thank you for this porting.

I have created a very simple form with C# .NET 7:

![image](https://user-images.githubusercontent.com/11708344/214822084-dd911e12-81ea-4974-a77c-a5f3d335a9a7.png)

WinForms.DataVisualization is at version 1.7.0 (Latest stable by the NuGet manager).

When I run this application on standard DPI (96, 100% scale) screens it works as intended:

![image](https://user-images.githubusercontent.com/11708344/214822309-b1d194b2-8943-4338-98a8-be3c76986cb3.png)

However after I move the window to a different DPI (240, 250% scale), all labels stays small:

![image](https://user-images.githubusercontent.com/11708344/214822389-987f7d9c-b99e-4bb6-80fe-9ab1da1c3cb1.png)

To illustrate the problem better, I scaled down this last image back to 100% (please don't consider the borders, lines, etc, focus only to labels on legend, and around the chart):

![image](https://user-images.githubusercontent.com/11708344/214822769-6b825ef3-a6fd-4860-8294-2996d44d2a95.png)

Can we somehow keep the consistency between DPI changes?

**Comment by kirsan31** [comment]
> First of all: thank you for this porting.

🙏

> Labels are not rescaled in higher DPI displays

I can't reproduce this behavior. Fresh app with `PerMonitorV2` enabled,  on the fly DPI changing results:
![image](https://user-images.githubusercontent.com/17767561/214905685-b5580616-5f97-4467-880a-880ae86b3a8a.png)

**Comment by danergo** [comment]
Thx for the quick heads up! I can confirm, works as intended if I change the monitor from x to y scale while application is running.

But that's not too real life scenario, i.e: I only change scale once per 10 years :)

However, if you have multiple monitors with different scale (100, 250), and you move the app between them, that's I would call a real life scenario.

And that scenario is not working. Lets move the app from 100% display to any other with different scaling.

**Comment by kirsan31** [comment]
WoW 😯It's very surprising...

What we have on this topic:
1. This is not chart problem, this is WinForms problem.
2. In theory changing settings on the fly must be equal to drugging app on other screen.
3. In practice, I heard that there were cases where dragging worked, but changing settings did not.
4. And now we have opposite behavior 🙄

Currently we have an interesting discussion on this topic [here](https://github.com/dotnet/winforms/issues/8376#issuecomment-1376028458) (read down from the linked post). You can join, or post an example application with this behavior here and I'll mention it there...

P.s. I have only one monitor therefore can test DPI related things only with settings change.

**Comment by danergo** [comment]
I kindof "fixed" this with using "DpiChangedBeforeParent" and "DpiChangedAfterParent" events.

**Comment by dreddy-work** [comment]
> I kindof "fixed" this with using "DpiChangedBeforeParent" and "DpiChangedAfterParent" events.

@danergo, Those events available for PermonV2 mode applications. And DataVisualization chart is ported only to support existing applications and we are not investing further to support higher DPI modes for this control. If labels here coming from that, explicit handling of those events is the right way to go.

**Comment by kirsan31** [comment]
@dreddy-work 
The main question here why it's working with on the fly dpi change then?

**Comment by danergo** [comment]
I found another very strange behavior: 

2 displays:
-250% (primary)
-100%

All fonts and lines (borders) MUST be manually rescaled to 100/250. I.e. font: 8.25F/2.5 to look correctly on the not-primary screen.

It seems as if the charting mechanism stores the primary Dpi, and it renders everything with that Dpi, and WinForms doesn't handle the Dpi change; SO in user code we have to deal with it (but in C# WinForms the Primary screen's Dpi value is not accessible).

**Comment by kirsan31** [comment]
Is this behavior the same with on the fly dpi change?
If _no_ (all working good), then we still need an answer from @dreddy-work...
And if _yes_, then I didn't get second part of your question 🙃

**Comment by danergo** [comment]
Similar!

On the fly dpi change I could solved by manually setting all fonts and borders to their inital size * DeviceDpi/96.

It worked perfectly, until a point my primary display is 100%. 
When my primary display is not 100%, setting a font to 8.25 * DeviceDpi/96 would make it extremely big on all screens.
If I don't messing with the manual sizing, and leave everything as-is, chart labels are fine on the primary screen only, all other screen they look either way too big or way too small.

**Comment by kirsan31** [comment]
Sorry man - still didn't get it :(

>  I can confirm, works as intended if I change the monitor from x to y scale while application is running.

vs

> On the fly dpi change I could solved by manually setting all fonts and borders to their inital size * DeviceDpi/96.

🤷‍♂️

let's try step by step what is not working - may me some video, or repro app (only on the fly change problems - I have no 2 monitor)?

**Comment by danergo** [comment]
No worries, I'm going to summarize here all my findings.

**Prerequisites:**

- 2 displays with different scale (i.e their Dpi is different)
- Compile the sample with DotNet SDK7
- Define `PerMonitorV2` in csproj
- WinForms.DataVisualization version 1.7.0

**ChartTest.csproj:**
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net7.0-windows</TargetFramework>
    <Nullable>enable</Nullable>
    <UseWindowsForms>true</UseWindowsForms>
    <ImplicitUsings>enable</ImplicitUsings>
    <ApplicationHighDpiMode>PerMonitorV2</ApplicationHighDpiMode>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="WinForms.DataVisualization" Version="1.7.0" />
  </ItemGroup>

</Project>
```

**Form1.cs:**
```
namespace ChartTest
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
    }
}
```

**Form1.Designer.cs:**
```
namespace ChartTest
{
    partial class Form1
    {
        /// <summary>
        ///  Required designer variable.
        /// </summary>
        private System.ComponentModel.IContainer components = null;

        /// <summary>
        ///  Clean up any resources being used.
        /// </summary>
        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code

        /// <summary>
        ///  Required method for Designer support - do not modify
        ///  the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            System.Windows.Forms.DataVisualization.Charting.ChartArea chartArea1 = new System.Windows.Forms.DataVisualization.Charting.ChartArea();
            System.Windows.Forms.DataVisualization.Charting.Legend legend1 = new System.Windows.Forms.DataVisualization.Charting.Legend();
            System.Windows.Forms.DataVisualization.Charting.Series series1 = new System.Windows.Forms.DataVisualization.Charting.Series();
            System.Windows.Forms.DataVisualization.Charting.DataPoint dataPoint1 = new System.Windows.Forms.DataVisualization.Charting.DataPoint(0D, 1D);
            System.Windows.Forms.DataVisualization.Charting.DataPoint dataPoint2 = new System.Windows.Forms.DataVisualization.Charting.DataPoint(0D, 4D);
            System.Windows.Forms.DataVisualization.Charting.DataPoint dataPoint3 = new System.Windows.Forms.DataVisualization.Charting.DataPoint(0D, 2D);
            this.chart1 = new System.Windows.Forms.DataVisualization.Charting.Chart();
            this.label1 = new System.Windows.Forms.Label();
            ((System.ComponentModel.ISupportInitialize)(this.chart1)).BeginInit();
            this.SuspendLayout();
            // 
            // chart1
            // 
            chartArea1.Name = "ChartArea1";
            this.chart1.ChartAreas.Add(chartArea1);
            legend1.Name = "Legend1";
            this.chart1.Legends.Add(legend1);
            this.chart1.Location = new System.Drawing.Point(12, 12);
            this.chart1.Name = "chart1";
            series1.ChartArea = "ChartArea1";
            series1.Legend = "Legend1";
            series1.Name = "Series1";
            series1.Points.Add(dataPoint1);
            series1.Points.Add(dataPoint2);
            series1.Points.Add(dataPoint3);
            this.chart1.Series.Add(series1);
            this.chart1.Size = new System.Drawing.Size(300, 300);
            this.chart1.TabIndex = 0;
            this.chart1.Text = "chart1";
            // 
            // label1
            // 
            this.label1.AutoSize = true;
            this.label1.Location = new System.Drawing.Point(318, 25);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(166, 15);
            this.label1.TabIndex = 1;
            this.label1.Text = "Series1 (Label for comparison)";
            // 
            // Form1
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(96F, 96F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Dpi;
            this.ClientSize = new System.Drawing.Size(527, 332);
            this.Controls.Add(this.label1);
            this.Controls.Add(this.chart1);
            this.Name = "Form1";
            this.Text = "Form1";
            ((System.ComponentModel.ISupportInitialize)(this.chart1)).EndInit();
            this.ResumeLayout(false);
            this.PerformLayout();

        }

        #endregion

        private System.Windows.Forms.DataVisualization.Charting.Chart chart1;
        private Label label1;
    }
}

```

This is how my designer looks like:

![image](https://user-images.githubusercontent.com/11708344/215978312-10abe1d9-3a0a-4121-8924-d18811f84150.png)

**Testcase 1:**

Displays:
![image](https://user-images.githubusercontent.com/11708344/215977677-5e410222-69bd-42e5-b0be-32746bbc221c.png)

I started 3 instances, and put one onto each display:

![image](https://user-images.githubusercontent.com/11708344/215978866-40a0c24d-faee-4d76-b6f6-ca6185906d4f.png)

As you can see, "Series1" label on the chart, on the 250% display is way smaller than the label for comparison, which initially has the same size (and on the primary screen it looks right).

This testcase proves that the Chart control is not able to rescale itself in case of Dpi changes due to moving it to a different display (with different Dpi).

Therefore, I have made a fix into Form's DpiChanged event:

```
private void Form1_DpiChanged(object sender, DpiChangedEventArgs e)
{
    chart1.Legends[0].Font = new System.Drawing.Font("Microsoft Sans Serif", 8.25F * e.DeviceDpiNew / 96, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point);
}
```

**Testcase 2:**
With the fix above, I have made a same screenshot with 3 instances:

![image](https://user-images.githubusercontent.com/11708344/215980309-ff2ac0b1-2f1a-4785-b96b-97762288709b.png)

You can see, now the Legend's text's size is matching with the test label's size. Please note also, that other texts on chart are still small.

Now, let's do some further testing.

**Testcase 3:**
Let's start the instances on the 250% display (Number1 - it's still not the primary!):

![image](https://user-images.githubusercontent.com/11708344/215981742-075d4e32-9b16-457b-a0ce-6b3969cca6d7.png)

Please note: instances from left to right:

- instance1: looks OK (as Form's DpiChange event fixes the font)
- instance2: looks OK (as Form's DpiChange event fixes the font)
- instance3: looks BAD (as we start the instances on this screen, there is no DpiChange event happened on this instance)

Okay, so let's call my "font fixer" in the Form's OnLoad too:

```
private void Form1_Load(object sender, EventArgs e)
{
    chart1.Legends[0].Font = new System.Drawing.Font("Microsoft Sans Serif", 8.25F * DeviceDpi / 96, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point);
}
```

**Testcase 4:**
Retesting with the above fix:

![image](https://user-images.githubusercontent.com/11708344/215982749-e8ee4cf8-1b80-44db-b3f3-4832d797d018.png)

All looks great now.

Let's do some further testing with this.

**Testcase 5:**

Displays:

![image](https://user-images.githubusercontent.com/11708344/215980952-8c3feef6-82c9-4830-859d-5176178ed80e.png)

Let's start the instances on the primary screen, and move place them the same as before:

![image](https://user-images.githubusercontent.com/11708344/215983390-8cb579e2-569c-49f4-947d-bb9e890c0ca8.png)

You can see, that legend's font is now oversized, BUT other texts on the chart looks good (ONLY on primary screen!)

Result: my font fixer doesn't work if the primary screen is not 100%. It seems the font size change to `DeviceDpi / 96` is not sufficient, because font size need to be set by taking into account also the primary screen's Dpi.

So after a crazy amount of digging for the information for the primary screen's Dpi, I changed font fixer to:

    chart1.Legends[0].Font = new System.Drawing.Font("Microsoft Sans Serif", 8.25F * DeviceDpi / PrimaryScreenDpi, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point);

And retested testcase4, and testcase5:

** Testcase 4 retest:**
![image](https://user-images.githubusercontent.com/11708344/215985268-02736b17-f203-49aa-aa08-546fbf8c8698.png)


** Testcase 5 retest:**
![image](https://user-images.githubusercontent.com/11708344/215984966-5a22c926-4ec3-486f-955e-6841a4524f78.png)

These are my findings, and unfortunately this also affects border sizes on both the chart axes and on linechart lines. Technically all line widths has to be corrected the same way as the fontsize I've shown above.

So, the correct scaling is **Current_Dpi_which_displays_the_chart / Primary_screens_dpi**.

Please note: none of these testing involved Dpi change or Primary display change during runtime. These changes were only done when the application was not running.

I hope now it's a bit clearer, let me know if there is anything unclear parts still. :)
Thank you!

**Comment by kirsan31** [comment]
@danergo
Came to say thank you for the detailed report. As soon as I have time, I will definitely try to figure it out...

**Comment by kirsan31** [comment]
@danergo

Ok, I've spend some time to to dig into this...
And found that chart control simple not support `PerMonitorV2` at all 😭
In [this test](https://github.com/kirsan31/winforms-datavisualization/issues/26#issuecomment-1405353600) I set up `PerMonitorV2` in project file, but forgot to call `ApplicationConfiguration.Initialize();` 🤦‍♂️So app was running in `SystemAware` mode. And this increased font is simple zooming...

That is. Only `SystemAware` out of the box, o manually support `PerMonitorV2`, like you tried, until  `PerMonitorV2` will be implements internally (I hope it will).

**Comment by kirsan31** [comment]
@danergo 
> I kindof "fixed" this with using "DpiChangedBeforeParent" and "DpiChangedAfterParent" events.

Hi, I wander - how you was able to _fix_ this?
As I can see, that all scaling done around `Graphics.DpiX/DpiY` and `Graphics` doesn't support `PerMonitorV2` - `DpiX` always return app start dpi.

----UPD----

Oh, I see - your mostly talked about fonts. Fonts are one side of the problem, we also need to scale graphics...

**Comment by kirsan31** [comment]
Implemented in [1.10 preview](https://github.com/kirsan31/winforms-datavisualization/releases/tag/1.10pr).
